File: backend/summarize_youtube.rb
require 'net/http'
require 'uri'
require 'json'
require 'rexml/document'
require 'cgi'
require 'readline'

def extract_caption_url(response_data)
  begin
    # استخراج بيانات ytInitialPlayerResponse
    regex = /var ytInitialPlayerResponse = (\{.*?\});/m
    match = response_data.match(regex)

    unless match
      # محاولة نمط بديل
      alt_regex = /ytInitialPlayerResponse\s*=\s*(\{.*?\});/m
      match = response_data.match(alt_regex)
    end

    if match && match[1]
      json_string = match[1]
      player_response = JSON.parse(json_string)

      caption_tracks = player_response.dig('captions', 'playerCaptionsTracklistRenderer', 'captionTracks')
      if caption_tracks && !caption_tracks.empty?
        base_url = caption_tracks[0]['baseUrl']
        # فك ترميز الحروف unicode
        base_url.gsub('\\u0026', '&')
      else
        puts 'No caption tracks found.'
        nil
      end
    else
      puts 'No caption data found in the response.'
      nil
    end
  rescue => e
    puts "An error occurred while extracting the caption URL: #{e.message}"
    nil
  end
end

def extract_video_id(input)
  regex = /(?:v=|\/)([0-9A-Za-z_-]{11})/
  match = input.match(regex)
  if match && match[1]
    match[1]
  elsif /^[0-9A-Za-z_-]{11}$/.match?(input)
    input
  else
    nil
  end
end

def fetch_transcript(caption_url)
  begin
    response = HTTParty.get(caption_url)
    
    if response.success?
      xml_data = response.body
      document = REXML::Document.new(xml_data)
      
      transcript = document.elements.collect('transcript/text') do |element|
        CGI.unescapeHTML(element.text || '')
      end
      
      transcript.join(' ')
    else
      STDERR.puts "Failed to fetch transcript. Status: #{response.code}"
      nil
    end
  rescue => e
    STDERR.puts "Error fetching or parsing the transcript: #{e.message}"
    nil
  end
end


def main
  print 'Enter YouTube video URL or ID: '
  user_input = gets.chomp
  video_id = extract_video_id(user_input)

  unless video_id
    puts 'Invalid YouTube video ID.'
    return
  end

  url = "https://www.youtube.com/watch?v=#{video_id}"
  begin
    uri = URI.parse(url)
    response = Net::HTTP.get_response(uri)
    response_data = response.body

    # استخراج رابط الترجمة
    caption_url = extract_caption_url(response_data)
    if caption_url
      # جلب النصوص
      transcript_text = fetch_transcript(caption_url)
      if transcript_text
        puts transcript_text
      else
        puts 'Failed to extract transcript text.'
      end
    else
      puts 'Failed to extract caption URL.'
    end
  rescue => e
    puts "Error fetching video data: #{e.message}"
  end
end

main()

File: backend/summarize_youtube.js
const axios = require('axios');
const fs = require('fs');
const prompt = require('prompt-sync')();
const { parseStringPromise } = require('xml2js');
const he = require('he');  // Import the 'he' library

function extractCaptionUrl(responseData) {
    try {
        // Extract ytInitialPlayerResponse JSON data
        const regex = /var ytInitialPlayerResponse = (\{.*?\});/s;
        let match = responseData.match(regex);

        if (!match) {
            // Try alternative pattern
            const altRegex = /ytInitialPlayerResponse\s*=\s*(\{.*?\});/s;
            match = responseData.match(altRegex);
        }

        if (match && match[1]) {
            const jsonString = match[1];
            const playerResponse = JSON.parse(jsonString);

            const captionTracks = playerResponse.captions?.playerCaptionsTracklistRenderer?.captionTracks;
            if (captionTracks && captionTracks.length > 0) {
                // You can choose the desired language here, defaulting to the first track
                let baseUrl = captionTracks[0].baseUrl;
                // Decode any unicode characters
                baseUrl = baseUrl.replace(/\\u0026/g, '&');
                return baseUrl;
            } else {
                console.log('No caption tracks found.');
                return null;
            }
        } else {
            console.log('No caption data found in the response.');
            return null;
        }
    } catch (error) {
        console.error('An error occurred while extracting the caption URL:', error);
        return null;
    }
}

function extractVideoId(input) {
    const regex = /(?:v=|\/)([0-9A-Za-z_-]{11})/;
    const match = input.match(regex);
    if (match && match[1]) {
        return match[1];
    } else if (/^[0-9A-Za-z_-]{11}$/.test(input)) {
        return input;
    } else {
        return null;
    }
}

async function fetchTranscript(captionUrl) {
    try {
        const response = await axios.get(captionUrl);
        const xmlData = response.data;

        // Parse the XML and extract the text content
        const result = await parseStringPromise(xmlData, { explicitArray: false });
        const transcript = result.transcript.text;

        if (Array.isArray(transcript)) {
            // Combine all text elements and decode HTML entities
            const transcriptText = transcript.map(item => he.decode(item._)).join(' ');
            return transcriptText;
        } else if (typeof transcript === 'object' && transcript._) {
            return he.decode(transcript._);
        } else {
            console.log("No transcript text found.");
            return null;
        }
    } catch (error) {
        console.error('Error fetching or parsing the transcript:', error.message);
        return null;
    }
}

(async function() {
    const userInput = prompt("Enter YouTube video URL or ID: ");
    const videoId = extractVideoId(userInput);

    if (!videoId) {
        console.log("Invalid YouTube video ID.");
        return;
    }

    const url = `https://www.youtube.com/watch?v=${videoId}`;
    try {
        const response = await axios.get(url, {
            headers: {
                'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko)',
                'Accept-Language': 'en-US,en;q=0.9',
            }
        });
        const responseData = response.data;

        // // Save the response to an HTML file for inspection
        // fs.writeFileSync("youtube_response.html", responseData, "utf8");
        // console.log("Response saved to youtube_response.html");

        const captionUrl = extractCaptionUrl(responseData);
        if (captionUrl) {
            // Fetch the transcript
            const transcriptText = await fetchTranscript(captionUrl);
            if (transcriptText) {
                console.log(transcriptText);
            } else {
                console.log("Failed to extract transcript text.");
            }
        } else {
            console.log("Failed to extract caption URL.");
        }
    } catch (error) {
        console.error("Error fetching video data:", error.message);
    }
})();